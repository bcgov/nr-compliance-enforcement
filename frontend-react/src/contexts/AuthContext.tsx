import { KeycloakLoginOptions } from 'keycloak-js';
import React, {
  createContext,
  useContext,
  useMemo,
  useState
} from 'react';
import KeycloakService from '../service/KeycloakService';
import KeycloakUser from '../types/KeycloakUser';

interface AuthContextData {
  signed: boolean;
  user: KeycloakUser | {};
  startKeycloak(): Promise<boolean>;
  login(options?: KeycloakLoginOptions): Promise<void>;
  logout(): Promise<void>;
  createLoginUrl(options?: KeycloakLoginOptions): string;
  provider: string;
  token: string | undefined,
}

const AuthContext = createContext<AuthContextData>({} as AuthContextData);

interface Props {
  children: React.ReactNode;
}

const AuthProvider: React.FC<{ children: React.ReactNode }> = ({ children }: Props) => {
  const [signed, setSigned] = useState<boolean>(false);
  const [user, setUser] = useState<KeycloakUser | {}>({});

  /**
   * Starts Keycloak instance.
   */
  async function startKeycloak() {
    try {
      const userIsLoggedIn = await KeycloakService.initKeycloak();
      setSigned(userIsLoggedIn);
      const kcUser = KeycloakService.getUser();
      setUser(kcUser);
      return userIsLoggedIn;
    } catch (e) {
      // eslint-disable-next-line no-console
      console.log('Keycloak init error:', e);
    }
    return false;
  }

  /**
   * Logout the user
   */
  async function logout() {
    try {
      await KeycloakService.logout();
      setUser({});
      setSigned(false);
    } catch (e) {
      // eslint-disable-next-line no-console
      console.error('Keycloak logout error:', e);
    }
  }

  const { createLoginUrl, login } = KeycloakService;
  const provider = KeycloakService.authMethod();
  const token = KeycloakService.getToken();

  // memoize
  const contextValue = useMemo(() => ({
    signed,
    user,
    startKeycloak,
    login,
    logout,
    createLoginUrl,
    provider,
    token
  }), [signed, user, startKeycloak, login, logout, createLoginUrl, provider, token]);

  return (
    <AuthContext.Provider value={contextValue}>
      { children }
    </AuthContext.Provider>
  );
};

/**
 * Create useAuth hook.
 *
 * @returns {AuthContext} context.
 */
function useAuth(): AuthContextData {
  const context = useContext(AuthContext);

  if (!context) {
    throw new Error('useAuth must be used within an AuthProvider');
  }

  return context;
}

export { AuthProvider, useAuth };
