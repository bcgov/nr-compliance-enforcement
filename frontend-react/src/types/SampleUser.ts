type SampleUser = {
  key?: number,
  firstName: string,
  lastName: string
}

export default SampleUser;
