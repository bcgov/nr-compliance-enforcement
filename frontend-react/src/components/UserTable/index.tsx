import React from 'react';

import {
  Table,
  TableHead,
  TableHeader,
  TableRow,
  TableBody,
  TableCell
} from '@carbon/react';
import { TrashCan } from '@carbon/icons-react';

import LoadingButton from '../LoadingButton';
import SampleUser from '../../types/SampleUser';
import { useAuth } from '../../contexts/AuthContext';

interface TableProps {
  elements: SampleUser[],
  deleteFn: Function,
  headers: string[]
}

const UserTable = ({ elements, deleteFn, headers }: TableProps) => {
  const { user } = useAuth();

  const loadingStatus = {
    loading: 'Deleting...',
    success: 'Deleted!',
    error: 'Error'
  };

  const hasWriteRole = (): boolean => {
    if ('roles' in user) {
      // 'user_write' is the name of the required rule
      return user.roles.includes('user_write');
    }
    return false;
  };

  const hashObject = (str: string): number => {
    return str.split('').reduce((s, c) => Math.imul(31, s) + c.charCodeAt(0) | 0, 0);
  };

  return (
    <Table size="lg" useZebraStyles={false}>
      <TableHead>
        <TableRow>
          {headers.map((header, idx) => (
            <TableHeader key={header} id={`header-${header}-${idx}`} data-testid={`header-${header}-${idx}`}>
              {header}
            </TableHeader>
          ))}
        </TableRow>
      </TableHead>
      <TableBody>
        {elements.map((item, idx) => (
          <TableRow key={hashObject(JSON.stringify(item))} id={`row${idx}`}>
            <TableCell>{idx}</TableCell>
            <TableCell>{item.firstName}</TableCell>
            <TableCell>{item.lastName}</TableCell>
            <TableCell>
              {!hasWriteRole()
                ? 'Some Text' : (
                  <LoadingButton
                    id={`delete-${idx}`}
                    clickFn={() => deleteFn(idx)}
                    label="Delete"
                    status={loadingStatus}
                    icon={TrashCan}
                  />
                )}
            </TableCell>
          </TableRow>
        ))}
      </TableBody>
    </Table>
  );
};

export default UserTable;
