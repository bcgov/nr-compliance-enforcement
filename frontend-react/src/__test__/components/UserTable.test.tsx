/* eslint-disable no-undef */
import React from 'react';

import { render } from '@testing-library/react';
import renderer from 'react-test-renderer';
import UserTable from '../../components/UserTable';
import SampleUser from '../../types/SampleUser';
import { AuthProvider } from '../../contexts/AuthContext';

describe('the UserTable component', () => {
  const users: SampleUser[] = [{
    firstName: 'User1FirstName',
    lastName: 'User1LastName'
  }, {
    firstName: 'User2FirstName',
    lastName: 'User2LastName'
  }];
  const tableHeaders: string[] = ['Column1', 'Column2', 'Column3'];

  it('should have the correct headers', () => {
    const { getByTestId } = render(
      <AuthProvider>
        <UserTable elements={users} deleteFn={() => {}} headers={tableHeaders} />
      </AuthProvider>
    );
    tableHeaders.forEach((element, i) => {
      expect(getByTestId(`header-${element}-${i}`).textContent).toBe(element);
    });
  });

  it('should match the snapshot', () => {
    const tree = renderer
      .create(
        <AuthProvider>
          <UserTable elements={users} deleteFn={() => {}} headers={tableHeaders} />
        </AuthProvider>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
