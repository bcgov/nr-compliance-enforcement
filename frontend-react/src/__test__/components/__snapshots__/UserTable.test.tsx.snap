// Vitest Snapshot v1

exports[`the UserTable component > should match the snapshot 1`] = `
<div
  className="cds--data-table-content"
>
  <table
    aria-describedby={null}
    aria-labelledby={null}
    className="cds--data-table cds--data-table--lg"
  >
    <thead>
      <tr>
        <th
          data-testid="header-Column1-0"
          id="header-Column1-0"
          scope="col"
        >
          <div
            className="cds--table-header-label"
          >
            Column1
          </div>
        </th>
        <th
          data-testid="header-Column2-1"
          id="header-Column2-1"
          scope="col"
        >
          <div
            className="cds--table-header-label"
          >
            Column2
          </div>
        </th>
        <th
          data-testid="header-Column3-2"
          id="header-Column3-2"
          scope="col"
        >
          <div
            className="cds--table-header-label"
          >
            Column3
          </div>
        </th>
      </tr>
    </thead>
    <tbody
      aria-live="polite"
    >
      <tr
        id="row0"
      >
        <td>
          0
        </td>
        <td>
          User1FirstName
        </td>
        <td>
          User1LastName
        </td>
        <td>
          Some Text
        </td>
      </tr>
      <tr
        id="row1"
      >
        <td>
          1
        </td>
        <td>
          User2FirstName
        </td>
        <td>
          User2LastName
        </td>
        <td>
          Some Text
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;
